### PR TITLE
refactor: use pl.stringx.rfind to replace the get_last_index func

### PR DESCRIPTION
**Pull request by monkeyDluffy6017**
### Description

We use the `get_last_index` function to get the last index of substring, but the function is not correct now.https://github.com/apache/apisix/blob/3cac54c84661fcb60416afe0234dcb0115a77370/apisix/plugins/log-rotate.lua#L132

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by tzssangglass** [comment]
can we add test cases to cover this?

**Comment by monkeyDluffy6017** [comment]
@tzssangglass  I can test the function with unit test, how can i export the local function `get_last_index`?

**Comment by tzssangglass** [comment]
> I can test the function with unit test, how can i export the local function `get_last_index`?

core.log.debug?

In fact, without testing, and with all the current test cases passing, I don't know where the original error was.

**Comment by monkeyDluffy6017** [comment]
How do I write a test case for this? The original function is faulty, but the final function happens to be fine
@spacewander

**Comment by spacewander** [comment]
> How do I write a test case for this? The original function is faulty, but the result happens to be fine @spacewander

We can provide some fake access log to see if the function can filter out the right one from the wrong ones?

**Comment by monkeyDluffy6017** [comment]
> get_last_index

Can I export the `get_last_index` function and do unit test?

**Comment by spacewander** [comment]
> Can I export the `get_last_index` function and do unit test?

Is this the only way?

**Comment by monkeyDluffy6017** [comment]
> Is this the only way?

It doesn't matter if we don't change it, the log-rotate plugin is now working fine, but someone should be careful if they are going to use the `get_last_index` function later

**Comment by spacewander** [comment]
> > Is this the only way?
> 
> It doesn't matter if we don't change it, the log-rotate plugin is now working fine, but someone should be careful if they are going to use the `get_last_index` function later

We can put it in the core.utils so it's able to test the function?

**Comment by monkeyDluffy6017** [comment]
@spacewander done

File: apisix/core/utils.lua [hunk 1]
```diff
@@ -335,4 +337,19 @@ end
 _M.resolve_var = resolve_var
 
 
+function _M.get_last_index(str, key)
+    local str_rev = str_reverse(str)
+    local key_rev = str_reverse(key)
+    local idx, _ = str_find(str_rev, key_rev)
```
Comment (spacewander) [inline apisix/core/utils.lua:343]: We can use plain string find here?

Comment (monkeyDluffy6017) [reply apisix/core/utils.lua]: Done

File: apisix/core/utils.lua [hunk 2]
```diff
@@ -335,4 +337,19 @@ end
 _M.resolve_var = resolve_var
 
 
+function _M.get_last_index(str, key)
+    local str_rev = str_reverse(str)
+    local key_rev = str_reverse(key)
+    local idx, _ = str_find(str_rev, key_rev, 1, true)
```
Comment (spacewander) [inline apisix/core/utils.lua:343]: We can use core_str.find for the plain str find

**Comment by spacewander** [comment]
Please make the CI pass, thanks!

**Review by spacewander — DISMISSED** [review]

**Review by kingluo — DISMISSED** [review]

**Review by soulbird — DISMISSED** [review]

File: apisix/core/utils.lua [hunk 3]
```diff
@@ -335,6 +336,21 @@ end
 _M.resolve_var = resolve_var
 
 
+function _M.get_last_index(str, key)
+    local str_rev = str_reverse(str)
+    local key_rev = str_reverse(key)
+    local idx, _ = core_str.find(str_rev, key_rev)
```
Comment (soulbird) [inline apisix/core/utils.lua:342]: `local idx = core_str.find(str_rev, key_rev)`

Comment (monkeyDluffy6017) [reply apisix/core/utils.lua]: Done

File: apisix/core/utils.lua [hunk 4]
```diff
@@ -335,6 +336,21 @@ end
 _M.resolve_var = resolve_var
 
 
+function _M.get_last_index(str, key)
```
Comment (kingluo) [inline apisix/core/utils.lua:339]: for such a basic string operation, I suggest reusing the `pl.stringx` from Penlight library. After all, we already import this lib in our code base. Never reinvent the wheel unless you have to.

https://lunarmodules.github.io/Penlight/libraries/pl.stringx.html#rfind

Example:

```
> print(require("pl.stringx").rfind("apisix__1928", "__"))
7
```

Comment (monkeyDluffy6017) [reply apisix/core/utils.lua]: Done

**Review by kingluo — APPROVED** [review]

**Pull request merged**